packages/lesswrong/components/ea-forum/giving-portal/AdminElectionVotes.tsx

```diff
@@ -0,0 +1,220 @@
+import sum from "lodash/sum";
+import { eaGivingSeason23ElectionName } from "../../../lib/eaGivingSeason";
+
+const styles = (theme: ThemeType) => ({
+  root: {
+    fontFamily: theme.palette.fonts.sansSerifStack,
+    fontSize: 14,
+    display: "flex",
+    flexDirection: "column",
+    gap: "16px",
+  },
+  numWinners: {
+    display: "flex",
+    gap: "16px",
+  },
+  table: {
+    borderCollapse: "collapse",
+    "& td, th": {
+      padding: 8,
+      textAlign: "left",
+    },
+  },
+});
+
+function approximatelyEqual(a: number, b: number, fractionalError = 1e-6) {
+  return Math.abs(a - b) < Math.max(Math.abs(a), Math.abs(b)) * fractionalError;
+}
+
+const normaliseVotes = (
+  votes: Record<string, number | null>[],
+  remainingCandidates: ElectionCandidateBasicInfo[]
+): Record<string, number>[] => {
+  const uniformVote = remainingCandidates.reduce((acc, { _id }) => {
+    acc[_id] = 1 / remainingCandidates.length;
+    return acc;
+  }, {} as Record<string, number>);
+
+  const normalisedVotes = votes.map((vote) => {
+    const totalValue = sum(Object.values(vote).filter((val) => val !== null)) ?? 0;
+
+    // If all the candidates they voted for have been eliminated, give them a vote that
+    // is uniform across all candidates
+    if (!totalValue) return uniformVote;
+
+    const normalisedVote = Object.entries(vote).reduce((acc, [candidate, val]) => {
+      if (!val) return acc; // Filter out null or zero values
+
+      acc[candidate] = val / totalValue;
+      return acc;
+    }, {} as Record<string, number>);
+
+    return normalisedVote;
+  });
+
+  // Assert:
+  // 1. The vote has been normalised to 1
+  // 2. No votes have been added or removed
+  if (normalisedVotes.length !== votes.length) {
+    throw new Error("Votes have been added or removed");
+  }
+
+  for (const vote of normalisedVotes) {
+    const totalValue = sum(Object.values(vote));
+
+    if (!approximatelyEqual(totalValue, 1)) {
+      throw new Error(`Vote has not been normalised to 1: ${JSON.stringify(vote)}`);
+    }
+  }
+
+  return normalisedVotes;
+};
+
+const aggregateVotes = (votes: Record<string, number>[]): Record<string, number> => {
+  const aggregatedVote = votes.reduce((acc, vote) => {
+    Object.entries(vote).forEach(([candidate, val]) => {
+      acc[candidate] = (acc[candidate] ?? 0) + val;
+    });
+    return acc;
+  }, {} as Record<string, number>);
+
+  // Assert:
+  // 1. The sum of the aggregated vote is equal to the number of votes
+  const totalValue = sum(Object.values(aggregatedVote));
+  if (!approximatelyEqual(totalValue, votes.length)) {
+    throw new Error(`Vote has not been aggregated correctly: ${JSON.stringify(aggregatedVote)}`);
+  }
+
+  return aggregatedVote;
+};
+
+const getSortedWinners = (
+  votes: Record<string, number | null>[],
+  candidates: ElectionCandidateBasicInfo[],
+  numWinners: number
+) => {
+  let votesWithEliminatedCandidates = [...votes];
+  let remainingCandidates = [...candidates];
```

Comment:
As a rule it's bad to mutate arguments that are passed in (unless that's expected). In this case I think copying them is not strictly necessary because they are just reassigned below rather than mutated, but I'm being extra careful